common/src/main/java/io/pravega/common/concurrent/ThreadPoolScheduledExecutorService.java

```diff
@@ -0,0 +1,388 @@
+/**
+ * Copyright Pravega Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.pravega.common.concurrent;
+
+import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * An implementation of {@link ScheduledExecutorService} which uses a thread pool. 
+ * 
+ * This class is similar to ScheduledThreadPoolExecutor but differs in the following ways:
+ * 
+ * 1. The thread pool supports growing. IE: {@code maxPoolSize} and {@code corePoolSize} don't have to be the same.
+ * 2. Queued tasks are stored in a lock-free queue so that no bottlenecks can occur on submit.
+ * 3. Scheduling a task without a delay is O(1) as opposed to O(log(n))
+ * 4. Canceling a task actually removes it from the queue and is O(n) as opposed to a no-op which leaves it in the queue or O(log(n)) when {@code setRemoveOnCancelPolicy(true)}.
```

Comment:
I wonder, could we avoid the `O(n)` if we mark the task as "cancelled" and just drop it when its time is up? I understand we may have a (temporary) leak of pointers, but it should avoid the `O(n)` task, wouldn't it?

Reply:
It is also holding onto anything in the stack of the runnable. Which may be non-trivial amount of memory. 
If it is considered desirable I could do this for non-delayed tasks, as these may be assumed to run almost immediately, but keep the removal logic for delayed ones.